### PR TITLE
Remove migrate borrowck mode

### LINKED ISSUE #43234: non-lexical lifetimes (NLL) tracking issue
This issue tracks the status of the transition to non-lexical lifetimes and a MIR-based borrow-checker. Both of these are jargon-y terms for compiler authors: the effect of these features on end-users is, simply put, that the compiler accepts a wider range of code with fewer bugs (and, hopefully, better error messages). We will refer to the combination of the above as NLL.

# Current status

Most of NLL is already enabled in current versions of Rust. However, 'migration mode' is still enabled (https://github.com/rust-lang/rust/issues/58781). This causes us to reject some code that we would eventually like to accept, and to use different code paths for emitting some diagnostics. The progress towards disabling 'migration mode' is tracked in https://github.com/rust-lang/rust/issues/58781

## Key facts for getting involved

- Weekly triage meeting **Wednesdays at 3:30pm Boston time** ([see in your timezone](http://www.thetimezoneconverter.com/?t=15:30&tz=Boston&))
  - Join the chat on [Zulip](https://rust-lang.zulipchat.com/)
  - Notes in [this dropbox paper document](https://paper.dropbox.com/doc/Non-lexical-lifetimes-NLL-Triage-Em2cJrvxQMMFWLE7lE5Be)
  - If you are unable to attend, minutes and further discussion takes place on [this internals thread](https://internals.rust-lang.org/t/lets-push-non-lexical-lifetimes-nll-over-the-finish-line/7115). 
  - Ask @nikomatsakis (on Zulip, ideally) to be added to the calendar invite.
- You can also find a [YouTube playlist of pair programming videos][yt]; these are sessions where we are working through how to solve a particular problem, often explaining big parts of the system on the way.

[yt]: https://www.youtube.com/playlist?list=PLCQVvhKUrTN9VfaQx2AyOSAstwENaGkYA

## Implementation plan

- [x] **Initial prototype:** features available on nightly builds with `#![feature(nll)]` for experimentation.
- [ ] **Valid code works:** (issues labeled with [NLL-complete])
  - [x] all run-pass tests pass
  - [x] bootstrap builds with nll enabled: see #51823.
     - summary: We have bootstrapped at least once. We need to integrate such bootstrapping into the CI to ensure it stops breaking. 
  - [x] crater run with nll enabled passes without regressions
    -  see #52217 for the set of ICEs from recent crater run.
- [ ] **Invalid code gets errors:** (issues labeled with [NLL-sound])
    - [ ] compile-fail/ui tests generate errors in the right places (possibly with bad diagnostics)
      - see #52663 as evidence that we generate errors in all cases where expected for the `ui` test suite. (Current plan for `compile-fail` suite is to port majority of those tests to `ui`)
- [ ] **Performance is good** (issues labeled with [NLL-performant])
    - see [this dropbox paper](https://paper.dropbox.com/doc/NLL-Performance-Tracking-Doc-dRlUMiWT8eexf80FTrYj8) which tracks profiling information
    - see also the [NLL perf dashboard](https://perf.rust-lang.org/nll-dashboard.html) which compares running `cargo check` without and with NLL turned on, so that we can see a summary of the current performance impact on the compiler's static analyses in isolation.
- [x] **Future compatibility warnings on stable:** Once crater runs are clean, we can enable the future-compatibility warnings on stable: see #46908 and #52681
  - [x] currently, due to #52681, we are issuing compatibility warnings on the 2018 edition.
  - [x] next step is to migrate the 2015 edition in the same manner (#57804)
- [ ] **Diagnostic parity:** (issues labeled with [NLL-diagnostics]) Ensure that all diagnostics are "as good or better" than before.
  - We will primarily validate this from our test base, but also consider errors that are reported "from the wild".
  - [x] round one: #49862
  - [x] round two: #52663
  - [ ] round three: #54528; remaining work items are cards on https://github.com/rust-lang/rust/projects/10
- [ ] transition from migration mode to full NLL
  - [x] NLL warnings (downgraded from errors) become errors again
  - [x] as an intermediate step, potentially use a deny-by-default lint before making these hard errors
- [ ] **Feature stabilized:** NLL and MIR-based borrow checker enabled by default,
- [x] AST-based borrow checker removed: https://github.com/rust-lang/rust/pull/64790.

Many of the work items for unchecked bullets above are gathered into one place on #57895

## How to join the working group

If you'd like to help, please **join the NLL working group** -- just leave a comment below or ping @nikomatsakis [on gitter](https://gitter.im/rust-impl-period/WG-compiler-nll) and you will be added to the @rust-lang/wg-compiler-nll team. You will then get occasional pings (e.g., when new mentoring instructions are available or when looking for help), as well as being eligible to be assigned to issues and so forth. We discuss things [on the WG-compiler-nll channel on Gitter](https://gitter.im/rust-impl-period/WG-compiler-nll). 

## How to find an issue to work on

To find important issues, use one of the following queries:

- Top priority right now are [issues tagged with `NLL-Complete`][NLL-complete], corresponding to code that should be accepted but isn't right now.
- Next priority are [issues tagged with `NLL-sound`][NLL-sound], corresponding to code that should get errors, but isn't right now.
- Next priority are [issues tagged with `NLL-diagnostics`][NLL-diagnostics], corresponding to low quality error messages.
- [Issues tagged with `NLL-performant`][NLL-performant], corresponding to cases where NLL analysis takes too long.

You may also wish to look for E-mentor issues, which means that they have mentoring instructions. Also, if an issue is assigned, then someone is supposed to be working on it, but it's worth checking if they have made progress and pinging them -- people often get busy with other things.

Issues tagged with [NLL-deferred] are low priority right now, but if one strikes your fancy, feel free to tackle it!

## How are issues are organized

All issues related to NLL are tagged with [`WG-compiler-nll`](). They are further tagged with a `NLL-foo` label to indicate a subcategory. Issues that have [no NLL-label] are considered "untriaged" and need to be sorted. Issues tagged with [NLL-deferred] are low priority right now.

[NLL-sound]: https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3ANLL-sound+
[NLL-complete]: https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3ANLL-complete+
[NLL-performant]: https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3ANLL-performant+
[NLL-diagnostics]: https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3ANLL-diagnostics+
[NLL-deferred]: https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3ANLL-deferred+
[NLL-fixed-by-NLL]: https://github.com/rust-lang/rust/labels/NLL-fixed-by-NLL
[`WG-compiler-nll`]: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AWG-compiler-nll
[no NLL-label]: https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3AWG-compiler-nll+-label%3ANLL-deferred+-label%3ANLL-performant+-label%3ANLL-diagnostics+-label%3ANLL-complete+-label%3ANLL-sound++-label%3ANLL-fixed-by-NLL

Finally, you can always take a look at the [full list of NLL-related issues](https://github.com/rust-lang/rust/labels/A-NLL).

In particular, issues tagged with `E-mentor` are those that contain mentoring instructions that can help you get started.

Issues (and pull requests) tagged with `I-nominated` are meant to be reviewed by the WG-compiler-nll at each weekly meeting. Here's the current [nominated list](https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=+label%3AI-nominated+label%3AA-NLL+).

If you can't find anything, [reach out to @nikomatsakis on gitter](https://gitter.im/rust-impl-period/WG-compiler-nll).

## Other issues

This section tracks related issues and notes. 

<details>
<summary>Bugs in AST borrow check fixed in MIR borrowck</summary>
  
https://github.com/rust-lang/rust/issues/47366 tracks known bugs from the AST borrow checker that were fixed in the MIR borrow checker. For each such bug, we have added a test to the repo with `#![feature(nll)]` and then closed the relevant issue (obviously, though, the bug will still be reproducable until the MIR-based borrow checker is stabilized, presuming one uses the AST-based borrow checker). You can also search for things tagged with [NLL-fixed-by-NLL].

</details>

<details>
  <summary>Questions to be resolved before stabilization</summary>
  
- [x] #46036: infinite loop false edges
- [x] NLL should identify and respect the lifetime annotations that the user wrote #47184
</details>

<details>
  <summary>Possible extensions</summary>
  
- [ ]  refine liveness with maybe-initialized at greater resolution than a single variable
</details>

### LINKED ISSUE #58781: NLL: turn off migration mode
Transition fully to NLL (instead of the migration mode currently used).

   * one currently opts into this via `#![feature(nll)]` or `-Z borrowck=mir`
   * It gets rid of AST-region inference errors, causing some code to be accepted that is rejected by migrate mode, and changing diagnostics in many cases.
   * Potential blockers: 
     * ~NLL-sound bugs that affect NLL mode without migration
       * ~#57642 (needs test for now)~
     * Diagnostic differences
       * List: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3ANLL-diagnostics
     * Migrate compare-mode=NLL test differences to revisions (to ease stabilization PR and meticulously check diagnostic differences)
     * ~Two-phase borrows (2PB) question: #56254~
     * ~Must have crater run showing no major crates fail in practice (#60680)~
     * #73154
       * Likely needs changes to the leak check
     * ~Investigate #96331~
       * Diagnostics-only change, should be fine

### PR DESCRIPTION
**Pull request by jackh726**
Closes #58781
Closes #43234 

# Stabilization proposal

This PR proposes the stabilization of `#![feature(nll)]` and the removal of `-Z borrowck`. Current borrow checking behavior of item bodies is currently done by first infering regions *lexically* and reporting any errors during HIR type checking. If there *are* any errors, then MIR borrowck (NLL) never occurs. If there *aren't* any errors, then MIR borrowck happens and any errors there would be reported. This PR removes the lexical region check of item bodies entirely and only uses MIR borrowck. Because MIR borrowck could never *not* be run for a compiled program, this should not break any programs. It does, however, change diagnostics significantly and allows a slightly larger set of programs to compile.

Tracking issue: #43234
RFC: https://github.com/rust-lang/rfcs/blob/master/text/2094-nll.md
Version: 1.63 (2022-06-30 => beta, 2022-08-11 => stable).

## Motivation

Over time, the Rust borrow checker has become "smarter" and thus allowed more programs to compile. There have been three different implementations: AST borrowck, MIR borrowck, and polonius (well, in progress). Additionally, there is the "lexical region resolver", which (roughly) solves the constraints generated through HIR typeck. It is not a full borrow checker, but does emit some errors.

The AST borrowck was the original implementation of the borrow checker and was part of the initially stabilized Rust 1.0. In mid 2017, work began to implement the current MIR borrow checker and that effort ompleted by the end of 2017, for the most part. During 2018, efforts were made to migrate away from the AST borrow checker to the MIR borrow checker - eventually culminating into "migrate" mode - where HIR typeck with lexical region resolving following by MIR borrow checking - being active by default in the 2018 edition.

In early 2019, migrate mode was turned on by default in the 2015 edition as well, but with MIR borrowck errors emitted as warnings. By late 2019, these warnings were upgraded to full errors. This was followed by the complete removal of the AST borrow checker.

In the period since, various errors emitted by the MIR borrow checker have been improved to the point that they are mostly the same or better than those emitted by the lexical region resolver.

While there do remain some degradations in errors (tracked under the [NLL-diagnostics tag](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3ANLL-diagnostics), those are sufficiently small and rare enough that increased flexibility of MIR borrow check-only is now a worthwhile tradeoff.

## What is stabilized

As said previously, this does not fundamentally change the landscape of accepted programs. However, there are a [few](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3ANLL-fixed-by-NLL) cases where programs can compile under `feature(nll)`, but not otherwise.

There are two notable patterns that are "fixed" by this stabilization. First, the `scoped_threads` feature, which is a continutation of a pre-1.0 API, can sometimes emit a [weird lifetime error](https://github.com/rust-lang/rust/issues/95527) without NLL. Second, actually seen in the standard library. In the `Extend` impl for `HashMap`, there is an implied bound of `K: 'a` that is available with NLL on but not without - this is utilized in the impl.

As mentioned before, there are a large number of diagnostic differences. Most of them are better, but some are worse. None are serious or happen often enough to need to block this PR. The biggest change is the loss of error code for a number of lifetime errors in favor of more general "lifetime may not live long enough" error. While this may *seem* bad, the former error codes were just attempts to somewhat-arbitrarily bin together lifetime errors of the same type; however, on paper, they end up being roughly the same with roughly the same kinds of solutions.

## What isn't stabilized

This PR does not completely remove the lexical region resolver. In the future, it may be possible to remove that (while still keeping HIR typeck) or to remove it together with HIR typeck.

## Tests

Many test outputs get updated by this PR. However, there are number of tests specifically geared towards NLL under `src/test/ui/nll`

## History

* On 2017-07-14, [tracking issue opened](https://github.com/rust-lang/rust/issues/43234)
* On 2017-07-20, [initial empty MIR pass added](https://github.com/rust-lang/rust/pull/43271)
* On 2017-08-29, [RFC opened](https://github.com/rust-lang/rfcs/pull/2094)
* On 2017-11-16, [Integrate MIR type-checker with NLL](https://github.com/rust-lang/rust/pull/45825)
* On 2017-12-20, [NLL feature complete](https://github.com/rust-lang/rust/pull/46862)
* On 2018-07-07, [Don't run AST borrowck on mir mode](https://github.com/rust-lang/rust/pull/52083)
* On 2018-07-27, [Add migrate mode](https://github.com/rust-lang/rust/pull/52681)
* On 2019-04-22, [Enable migrate mode on 2015 edition](https://github.com/rust-lang/rust/pull/59114)
* On 2019-08-26, [Don't downgrade errors on 2015 edition](https://github.com/rust-lang/rust/pull/64221)
* On 2019-08-27, [Remove AST borrowck](https://github.com/rust-lang/rust/pull/64790)

### CODE REVIEW

**Comment by jackh726** [comment]
@cratebot check

**Comment by compiler-errors** [comment]
assuming jack meant

@craterbot check

**Comment by jackh726** [comment]
I'm going to nominate for lang team. Particularly, I'd like to see the following discussed:
1) There are a number of nontrivial breakage found with crater. This is likely covered by #57895. Are we willing to accept this breakage, or do we somehow need to engineer a way to get around it in older editions?
2) General temperature on stabilizing NLL.

(I'm going to try to add a couple more points this weekend to discuss).

**Comment by m-ou-se** [comment]
> :tada: Experiment **`pr-95565`** is completed!
> :bar_chart:  454 regressed and 33 fixed (223389 total)
> :newspaper: [Open the full report](https://crater-reports.s3.amazonaws.com/pr-95565/index.html).


160 of the 179 root regressions that craterbot found previously already warned the user the code will stop working in the future:

```
warning: this borrowing pattern was not meant to be accepted, and may become a hard error in the future
```

The other 19 are:

(Found using `$ for c in $(jq -r '.crates[] | select(.res == "regressed") | select(.runs[1].res | startswith("build-fail:depends") | not).runs[0].log | "https://crater-reports.s3.amazonaws.com/pr-95565/" + . + "/log.txt"' results.json); do curl -s $c | grep -q 'may become a hard error in the future' || echo $c; done`)

- 12 "cannot borrow as mutable because it is also borrowed as immutable":
  - gh/Belovedbb.Crlox [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/Belovedbb.Crlox/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/Belovedbb.Crlox/log.txt)
  - gh/GCYYfun.OSTEP_RUST [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/GCYYfun.OSTEP_RUST/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/GCYYfun.OSTEP_RUST/log.txt)
  - gh/gjconklin1.advent2021 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/gjconklin1.advent2021/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/gjconklin1.advent2021/log.txt)
  - gh/jumang4423.nylang [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/jumang4423.nylang/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/jumang4423.nylang/log.txt)
  - gh/laopo001.leetcode-rust [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/laopo001.leetcode-rust/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/laopo001.leetcode-rust/log.txt)
  - gh/mwillsey.ruler [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/mwillsey.ruler/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/mwillsey.ruler/log.txt)
  - gh/qinxiaoguang.rs-lc [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/qinxiaoguang.rs-lc/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/qinxiaoguang.rs-lc/log.txt)
  - reg/ffsend-0.2.74 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/ffsend-0.2.74/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/ffsend-0.2.74/log.txt)
  - reg/notedown_ast-0.16.2 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/notedown_ast-0.16.2/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/notedown_ast-0.16.2/log.txt)
  - reg/rx-editor-0.3.0 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/rx-editor-0.3.0/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/rx-editor-0.3.0/log.txt)
  - reg/voltdb-client-rust-0.1.11 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/voltdb-client-rust-0.1.11/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/voltdb-client-rust-0.1.11/log.txt)
  - reg/wire4-0.0.2 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/wire4-0.0.2/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/wire4-0.0.2/log.txt)

- 7 spurious or unrelated:
  - gh/Ayrx.binja-rs-hello-world [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/Ayrx.binja-rs-hello-world/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/Ayrx.binja-rs-hello-world/log.txt)
  - gh/ababup1192.elm-init [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/ababup1192.elm-init/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/ababup1192.elm-init/log.txt)
  - gh/etke.bff [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/etke.bff/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/etke.bff/log.txt)
  - gh/lcdr.lu_packets [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/lcdr.lu_packets/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/lcdr.lu_packets/log.txt)
  - gh/stuartZhang.scaffold-wizard [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/gh/stuartZhang.scaffold-wizard/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/gh/stuartZhang.scaffold-wizard/log.txt)
  - reg/bitrust-0.1.0 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/bitrust-0.1.0/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/bitrust-0.1.0/log.txt)
  - reg/rttrust-0.1.3 [before](https://crater-reports.s3.amazonaws.com/pr-95565/master%23297a8018b525c28ef10ee6a91d61954839b508b9/reg/rttrust-0.1.3/log.txt) [after](https://crater-reports.s3.amazonaws.com/pr-95565/try%23eb43a7af9abeacf703345e49daba222ddc0623e9/reg/rttrust-0.1.3/log.txt)

**Comment by m-ou-se** [comment]
In every single one of the cases where the future compatibility warning didn't fire on code that's broken after this PR, the programmer manually supressed the warning.

**Comment by pnkfelix** [comment]
cc https://github.com/rust-lang/rust/pull/96268

**Comment by joshtriplett** [comment]
@rfcbot merge

@rfcbot concern wait-for-96268
@rfcbot concern confirm-no-other-regressions

**Comment by jackh726** [comment]
To also clarify, we haven't finished going through all the open diagnostic changes between NLL and migrate. It's already generated at least one difference that needs to be look at: #96331.

I've updated the tracking issue (#58781) to include these two and #73154.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

Approved modulo existing concerns!

**Comment by nikomatsakis** [comment]
@rfcbot concern investigate-96331

We should figure out whether #96331 represents expected behavior.

**Comment by joshtriplett** [comment]
@rfcbot resolved confirm-no-other-regressions

Got confirmation from @m-ou-se that all the root regressions arose from `mutable_borrow_reservation_conflict`, which we're removing.

**Comment by joshtriplett** [comment]
@nikomatsakis We discussed that concern in today's @rust-lang/lang meeting. Everyone seemed to feel that the diagnostics could be *better* (reporting more), but that this shouldn't be a blocker for NLL stabilization.

**Comment by nikomatsakis** [comment]
@rfcbot resolve  investigate-96331

Thanks @m-ou-se

**Comment by jackh726** [comment]
I have a PR for #73154 up (#97206). That should be the last "blocking" thing. Just need to go through the remaining tests with compare-mode=nll differences, convert to revisions, and triage.

**Comment by jackh726** [comment]
Status update: I've rebased this onto #97258 (which in turn is based on #97206). I've gone through and "converted" the NLL revision tests as they should be with this PR. There are ~60ish tests that aren't passing with the mechanical changes of just removing the revisions and appropriate error annotations. A lot of these fall into the directories listed in #97258, so those are expected. Going to look into the others though and see if there's anything I can do targeting the master branch to make this PR solely mechanical.

**Comment by jackh726** [comment]
I've added a stabilization report to the OP and rebased.

File: src/test/ui/borrowck/issue-55492-borrowck-migrate-scans-parents.migrate.stderr [hunk 1]
```diff
@@ -1,54 +1,8 @@
...
+error: expected identifier, found `<<`
+  --> $DIR/issue-55492-borrowck-migrate-scans-parents.rs:5:1
    |
-LL |     pub fn e(x: &'static mut isize) {
-   |              - help: consider changing this to be mutable: `mut x`
-LL |         static mut Y: isize = 3;
-LL |         let mut c1 = |y: &'static mut isize| x = y;
-   |                                              ^^^^^ cannot assign
+LL | <<<<<<< HEAD
+   | ^^ expected identifier
```
Comment (klensy) [inline src/test/ui/borrowck/issue-55492-borrowck-migrate-scans-parents.migrate.stderr:5]: git rebase leftovers?

File: src/test/ui/borrowck/two-phase-reservation-sharing-interference-2.migrate2015.stderr [hunk 2]
```diff
@@ -20,6 +20,22 @@ LL |     v.extend(&v);
    |     | immutable borrow later used by call
    |     mutable borrow occurs here
 
+<<<<<<< HEAD
 error: aborting due to 2 previous errors
+=======
```
Comment (klensy) [inline src/test/ui/borrowck/two-phase-reservation-sharing-interference-2.migrate2015.stderr:25]: and here

File: src/test/ui/borrowck/two-phase-reservation-sharing-interference-2.migrate2018.stderr [hunk 3]
```diff
@@ -20,6 +20,22 @@ LL |     v.extend(&v);
    |     | immutable borrow later used by call
    |     mutable borrow occurs here
 
+<<<<<<< HEAD
 error: aborting due to 2 previous errors
+=======
```
Comment (klensy) [inline src/test/ui/borrowck/two-phase-reservation-sharing-interference-2.migrate2018.stderr:25]: and here

File: src/test/ui/hrtb/issue-30786.migrate.stderr [hunk 4]
```diff
@@ -1,20 +1,21 @@
-error[E0599]: the method `filterx` exists for struct `Map<Repeat, [closure@$DIR/issue-30786.rs:121:27: 121:36]>`, but its trait bounds were not satisfied
-  --> $DIR/issue-30786.rs:122:22
+error: expected identifier, found `<<`
+  --> $DIR/issue-30786.rs:9:1
    |
+<<<<<<< HEAD
```
Comment (klensy) [inline src/test/ui/hrtb/issue-30786.migrate.stderr:4]: and here

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me -- but I suggested to @jackh726 that we ought to run a perf run, just because

**Comment by jackh726** [comment]
Cool. Okay, let's do this. 

@bors r=nikomatsakis

**Pull request merged**